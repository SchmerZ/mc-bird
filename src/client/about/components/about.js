import React, {Fragment} from 'react'
import styled from 'styled-components'

import Links from './links'
import Thanks from './thanks'
import UsedFrameworks from './used-farmeworks'
import PotentialFeatures from './potential-features'

const Title = styled.h3`
  margin-top: 0;
`;

const About = () => {
  return (
    <Fragment>
      <Title>mc-Bird (Vadim Sheydakov)</Title>
      <Links />
      <Thanks />
      <UsedFrameworks />
      <PotentialFeatures />
    </Fragment>
  )
};

export default About
