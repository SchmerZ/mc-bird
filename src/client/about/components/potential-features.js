import React, {Fragment} from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import {color} from '../../styles/variables'

const Title = styled.h4`
  margin-bottom: 0;
`;

const FeatureContainer = styled.div`
  padding: 4px 0;
`;

const FeatureName = styled.span`
  font-style: italic;
  font-weight: 600;
  color: ${color.primary};
`;

const FeatureRow = ({name, description}) => {
  return (
    <FeatureContainer>
      <FeatureName>{name}. </FeatureName>
      {description}
    </FeatureContainer>
  )
};

FeatureRow.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
};

const PotentialFeatures = () => {
  return (
    <Fragment>
      <Title>Potential features:</Title>
      <FeatureRow
        name="loadable components"
        description="Avoid page parts 'blinking' from spinner icon to content. If we are able to get data faster then specified threshold, no need to render a spinning icon for a few ms. Wait instead."
      />

      <FeatureRow
        name="websocket"
        description="Can be used for outcome messages as well. Since not having much experience, postponed for the future exploring."
      />

      <FeatureRow
        name="webpack code-splitting"
        description="Great tool to decrease initial js package size."
      />
    </Fragment>
  )
};

export default PotentialFeatures
