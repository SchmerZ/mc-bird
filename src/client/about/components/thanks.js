import React, {Fragment} from 'react'
import styled from 'styled-components'

const Title = styled.h4`
  margin-bottom: 0;
`;

const Paragraph = styled.p``;

const Thanks = () => {
  return (
    <Fragment>
      <Title>Thanks:</Title>
      <Paragraph>
        I would like to thank HR <strong>Chrisje Alsters</strong> for patience in chat and giving me a chance to try
        myself in that assignment.
      </Paragraph>

      <Paragraph>
        Some things (i.e. websocket) I tried for the first time because it fits to current solution. So I would appreciate any feedback about the code or technical decisions
        which take place.
      </Paragraph>

      <Paragraph>
        Latest version of <strong>MessageBird REST API npm client</strong> has limited support of functionality. So I have
        written own client from scratch.
      </Paragraph>

      <Paragraph>
        Great <strong>thanks</strong> for the reviewers who is responsible to check my code.
        Hope <strong>#?%!</strong> was not the most frequent word you pronounced =).
      </Paragraph>
    </Fragment>
  )
};

export default Thanks
