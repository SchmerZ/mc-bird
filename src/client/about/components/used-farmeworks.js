import React, {Fragment} from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import {color} from '../../styles/variables'
import {Row, Column} from '../../../shared/components/layout/responsive'

const Title = styled.h4`
  margin-bottom: 0;
`;

const FrameworkContainer = styled(Row)`
  display: flex;
  align-items: center;
  padding: 4px 0;
`;

const FrameworkColumn = styled(Column)`
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  color: ${color.primary};
`;

const FrameworkRow = ({name, description}) => {
  return (
    <FrameworkContainer>
      <FrameworkColumn xs={4} md={2}>
        {name}
      </FrameworkColumn>
      <Column xs={8} md={10}>
        {description}
      </Column>
    </FrameworkContainer>
  )
};

FrameworkRow.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
};

const UsedFrameworks = () => {
  return (
    <Fragment>
      <Title>Used frameworks:</Title>
      <FrameworkRow
        name="webpack"
        description="Transpile'n'pack js. Justify for production. Code splitting (not implemented). Writing js on fancy and geeky way."
      />

      <FrameworkRow
        name="websocket"
        description="Live update on income'n'outcome messages. Had no chance to try before. Would like to play with."
      />

      <FrameworkRow
        name="redux + saga"
        description="Redux helps to build scalable SPA with logic segregation. Saga is a powerful framework to handle side effects. Easy to maintain scenarios like polling/infinite loop action handlers."
      />

      <FrameworkRow
        name="reselect"
        description="'Computed' functions to cache result and recompute on references change. Helps to keep performance."
      />

      <FrameworkRow
        name="connected-react-router"
        description="Provides actions to work with browser history. Extend store with params for current browser location."
      />

      <FrameworkRow
        name="styled-components"
        description="Css-in-js. I don't have big experience with css-in-js paradigm. Has cons/pros (as any other). Have tried server side rendering and it works =)"
      />

      <FrameworkRow
        name="nconf"
        description="Hierarchical sources support for configuration (argv, environment, file). With validation and override support on each level."
      />

      <FrameworkRow
        name="jest + enzyme"
        description="Fancy tests for React components. Snapshot testing."
      />

      <FrameworkRow
        name="storybook"
        description="A place where it's good to start creating components (proof-of-concept). Initial test/demo for components."
      />
    </Fragment>
  )
};

export default UsedFrameworks
