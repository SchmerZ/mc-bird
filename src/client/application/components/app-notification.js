import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {bindActionCreators} from 'redux'
import {connect} from 'react-redux'

import * as A from '../actions'

import ConsecutiveSnackbars from '../../components/snackbars/consecutive-snackbars'

export class AppNotification extends Component {
  handleMessageExited = () => {
    this.props.messageClose();
  };

  render() {
    const {messages} = this.props;

    return (
      <ConsecutiveSnackbars
        messages={messages}
        onExited={this.handleMessageExited}
      />
    )
  }
}

AppNotification.propTypes = {
  messages: PropTypes.array,
  messageClose: PropTypes.func,
};

const mapStateToProps = (state) => {
  const {application: {messages}} = state;

  return {
    messages,
  }
};

const mapDispatchToProps = (dispatch) => bindActionCreators(A, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AppNotification)
