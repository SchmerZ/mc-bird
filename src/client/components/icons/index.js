import SendIcon from './send-icon'
import ChatIcon from './chat-icon'
import ConversationIcon from './conversation-icon'
import MessageIcon from './message-icon'
import SmsIcon from './sms-icon'
import SpinnerIcon from './spinner-icon'
import MessageBirdLogoIcon from './mb-logo-icon'
import MessageBirdHeaderLogoIcon from './mb-header-logo-icon'
import ProfileIcon from './profile-icon'
import Icon404 from './404'
import InfoIcon from './info-icon'
import ErrorIcon from './error-icon'
import CheckCircleIcon from './check-circle-icon'
import CloseIcon from './close-icon'

export {
  SendIcon,
  ChatIcon,
  SmsIcon,
  ConversationIcon,
  MessageIcon,
  SpinnerIcon,
  MessageBirdLogoIcon,
  MessageBirdHeaderLogoIcon,
  ProfileIcon,
  Icon404,
  InfoIcon,
  ErrorIcon,
  CheckCircleIcon,
  CloseIcon,
}
