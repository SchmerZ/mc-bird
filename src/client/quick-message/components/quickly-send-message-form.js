import React, {Component, Fragment} from 'react'
import PropTypes from 'prop-types'
import {bindActionCreators} from 'redux'
import {connect} from 'react-redux'
import styled from 'styled-components'

import * as A from '../actions'

import {Row, Column} from '../../components/layout/responsive'
import TextField from '../../components/text-field/text-field'
import SendMessageText from '../../components/message/send-message-text'
import SendMessageButton from './send-message-button'

const SendMessageTextColumn = styled(Column)`
  margin-bottom: 15px;
`;

const SendSmsButtonContainer = styled.div`
  text-align: right;
  padding-right: 15px;
`;

export class QuicklySendMessageForm extends Component {
  handleRecipientChange = (value) => {
    const {changeRecipient} = this.props;
    changeRecipient && changeRecipient(value);
  };

  handleMessageTextChange = (value) => {
    const {changeMessageText} = this.props;
    changeMessageText && changeMessageText(value);
  };

  handleSendMessageClick = () => {
    const {send} = this.props;
    send && send();
  };

  render() {
    const {recipient, messageText, sending} = this.props;

    return (
      <Fragment>
        <Row>
          <Column md={6}>
            <TextField
              debounceTimeout={300}
              placeholder="Recipient"
              defaultValue={recipient}
              onChange={this.handleRecipientChange}
            />
          </Column>
        </Row>
        <Row>
          <SendMessageTextColumn md={12}>
            <SendMessageText
              value={messageText}
              onChange={this.handleMessageTextChange}
            />
          </SendMessageTextColumn>
        </Row>
        <Row>
          <SendSmsButtonContainer>
            <SendMessageButton busy={sending} onClick={this.handleSendMessageClick}/>
          </SendSmsButtonContainer>
        </Row>
      </Fragment>
    )
  }
}

QuicklySendMessageForm.propTypes = {
  recipient: PropTypes.string,
  messageText: PropTypes.string,
  sending: PropTypes.bool,

  send: PropTypes.func,
  changeRecipient: PropTypes.func,
  changeMessageText: PropTypes.func,
};

const mapStateToProps = (state) => {
  const {
    quickMessage: {
      recipient,
      messageText,
      sending
    }
  } = state;

  return {
    sending,
    recipient,
    messageText,
  }
};

const mapDispatchToProps = (dispatch) => bindActionCreators(A, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(QuicklySendMessageForm)