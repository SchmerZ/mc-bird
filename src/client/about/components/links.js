import React, {Fragment} from 'react'
import styled from 'styled-components'

import {Anchor} from '../../components/styled/primitive'
import {GitHubIcon, LinkedInIcon} from '../../components/icons'

const Title = styled.h4`
  margin-bottom: 0;
`;

const LinkContainer = styled.div`
  display: flex;
  align-items: center;
  padding: 4px 0;
`;

const StyledAnchor = styled(Anchor)`
  margin-left: 10px;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const iconSize = 25;

const Links = () => {
  return (
    <Fragment>
      <Title>Links:</Title>
      <LinkContainer>
        <GitHubIcon size={iconSize} />
        <StyledAnchor target="__blank"
                      href="https://github.com/SchmerZ/mc-bird">https://github.com/SchmerZ/mc-bird</StyledAnchor>
      </LinkContainer>

      <LinkContainer>
        <LinkedInIcon size={iconSize} />
        <StyledAnchor target="__blank"
                      href="https://www.linkedin.com/in/vadimsheydakov/">https://www.linkedin.com/in/vadimsheydakov/</StyledAnchor>
      </LinkContainer>
    </Fragment>
  )
};

export default Links
