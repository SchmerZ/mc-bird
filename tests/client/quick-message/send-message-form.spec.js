import React from 'react';
import {shallow} from 'enzyme';

import {QuicklySendMessageForm} from '../../../src/client/quick-message/components/quickly-send-message-form';
import SendMessageButton from '../../../src/client/quick-message/components/send-message-button';

import TextField from '../../../src/client/components/text-field/text-field';
import SendMessageText from '../../../src/client/components/message/send-message-text';

describe('Quick message form', () => {
  describe('Recipient TextField', () => {
    it('should render with passed props', () => {
      const form = shallow(<QuicklySendMessageForm recipient="Recipient value"/>);
      const textField = form.find(TextField);

      expect(textField).toHaveLength(1);
      expect(textField.props()).toMatchSnapshot();
    });

    it('should pass error props if form has recipientError props', () => {
      const errors = {recipientError: 'some recipient error'};
      const form = shallow(<QuicklySendMessageForm recipient="Recipient value" errors={errors}/>);
      const textField = form.find(TextField);

      expect(textField.props().error).toEqual(errors.recipientError);
    });

    it('should rise changeRecipient props function if text has changed', () => {
      const fn = jest.fn();
      const form = shallow(<QuicklySendMessageForm changeRecipient={fn}/>);

      form.find(TextField).simulate('change', 'text');
      expect(fn).toBeCalledWith('text');
    });
  });

  describe('SendMessageText', () => {
    it('should render SendMessageText with passed props', () => {
      const form = shallow(<QuicklySendMessageForm messageText="Some message text"/>);
      const sendText = form.find(SendMessageText);

      expect(sendText).toHaveLength(1);
      expect(sendText.props()).toMatchSnapshot();
    });

    it('should pass error props if form has messageTextError props', () => {
      const errors = {messageTextError: 'some message error'};
      const form = shallow(<QuicklySendMessageForm messageText="Some message text" errors={errors}/>);
      const sendText = form.find(SendMessageText);

      expect(sendText.props().error).toEqual(errors.messageTextError);
    });

    it('should rise changeRecipient props function if text has changed', () => {
      const fn = jest.fn();
      const form = shallow(<QuicklySendMessageForm changeRecipient={fn}/>);

      form.find(TextField).simulate('change', 'text');
      expect(fn).toBeCalledWith('text');
    });
  });

  describe('SendMessage button', () => {
    it('should render SendMessageButton with passed props', () => {
      const form = shallow(<QuicklySendMessageForm sending/>);
      const button = form.find(SendMessageButton);

      expect(button).toHaveLength(1);
      expect(button.props()).toMatchSnapshot();
    });

    it('should call "send" props function on click', () => {
      const fn = jest.fn();
      const form = shallow(<QuicklySendMessageForm send={fn}/>);

      form.find(SendMessageButton).simulate('click');
      expect(fn).toBeCalledTimes(1);
    });
  });
});
