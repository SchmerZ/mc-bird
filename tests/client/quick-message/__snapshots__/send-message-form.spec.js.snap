// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quick message form Recipient TextField should render with passed props 1`] = `
Object {
  "debounceTimeout": 300,
  "defaultValue": "Recipient value",
  "error": undefined,
  "onChange": [Function],
  "placeholder": "Recipient",
}
`;

exports[`Quick message form SendMessage button should render SendMessageButton with passed props 1`] = `
Object {
  "busy": true,
  "onClick": [Function],
}
`;

exports[`Quick message form SendMessageText should render SendMessageText with passed props 1`] = `
Object {
  "error": undefined,
  "maxLength": 1377,
  "onChange": [Function],
  "value": "Some message text",
}
`;
